### PR TITLE
do not drop mouseover on WM_MOUSE_EXIT

### LINKED ISSUE #39634: Active scrollbar loses focus if mouse leaves window
**Godot version:**
Introduced in 5b00cd89d318a92e7970fbdf80d9fca424e15224

**OS/device including version:**
Linux

**Issue description:**
Practically every application I know of lets you leftclick a scrollbar and start dragging up and down, and if your mouse leaves the window you can still keep on scrolling as the scrollbar retains focus. This is very useful since scrollbars often are at the very edge of the window where the mouse is prone to exiting. 

Scrollbars randomly dropping focus because the mouse just barely left the window is incredibly infuriating.

**Steps to reproduce:**
* Start the engine/editor to reach the project manager screen
* Click the templates tab to load up some templates and make a scrollbar appear
* Click and hold left mousebutton on the scrollbar
* Drag the scrollbar up and down a bit
* Move the mouse outside the window while still holding the scrollbar
* The scrollbar drops focus and stops scrolling (although the window itself remains focused until the mousebutton is released)

I expect the scrollbar to not drop focus.

### PR DESCRIPTION
**Pull request by Meriipu**
closes  #39634

### CODE REVIEW

**Comment by Meriipu** [comment]
A new issue with this is that the scrollbar remains highlighted if the mouse is released while the cursor is outside the window. 
This might be an issue with my window manager which has focus that follows the mouse-cursor, so maybe there is a need to add a call somewhere that on focus lost, release (or something).

**Comment by Meriipu** [comment]
Perhaps it is only desirable to drop focus if the mouse was actually held as it exited? 

I have no idea. The mentioned "stays highlighted if mouse released outside window"-issue remains, although it is likely best handled separately as it also occurs if the mouse is released while in the middle of the window (both in this commit and in 5b00cd8, though that one clears it on wm_mouse_exit).

Again my window manager might also be part of this highlight sticking-behaviour, I have not tested other window managers with different focus behaviours.

**Comment by RandomShaper** [comment]
In my opinion, the ideal behavior would be this (in both cases, both for GUI and 2D/3D physics, where it applies):
- Hovering is dropped once the pointer exits the viewport.
- "Dragging" is dropped if the mouse button that was held to start it is released with the pointer outside the viewport.

In the second case. maybe it would be better to keep the hovering too if dragging is kept and releasing both in the usual order when the mouse button is released, either with the pointer inside or outside the viewport.

I'm trying to consider all relevant cases, including sub-viewports, smaller than the whole window.

What do you think?

**Comment by Meriipu** [comment]
Is hovering is used for anything while something is being dragged (visually or otherwise)? I notice that when it comes to scrollbars, both godot and various gtk-applications have one style for hovering and another for draggin, and I can not see any way to discern that hovering is dropped while dragging is active (there may well be underlying events triggered I guess).

What you suggest sounds good to me, though (either of the two). 

I am not sure how sub-viewports work, but do you mean that the dragging will drop if the mouse exits the sub-viewport (and enters an outer viewport)? Are there any cases where that would be preferable behaviour vs have dragging stay active no matter where the mouse is (inside or outside the whole wm-window, cross viewport-boundaries)? 

There are almost certainly use cases I have not taken into account.

**Comment by Meriipu** [comment]
I guess what I am asking is: is there any control or 2D/3D feature which starts functioning in undesirable ways on `wm_mouse_exit`  with this partial revert/change?
If it is just a minor inconvenience like staying cosmetically highlighted until the mouse reenters, then I would personally prefer that to the current behaviour. 

Of course, fixing the highlighting once the `mouse pressed`-state ends (before moving the mouse to update controls) is also desirable
but honestly this lingering highlighting appears to be a separate issue from the dropped scrolling, as it happens even when the mouse is inside the window but not moving (after release), both before and after this commit. 

So in short are there any issues with this fix?

**Comment by Meriipu** [comment]
Accidentally went one commit too far back in the first rebase. 
The log looked fine but apparently the one-too-far commit was rewritten to be some sort of co-author commit. 
It should be fixed now though.

**Comment by RandomShaper** [comment]
> So in short are there any issues with this fix?

If the user eventually releases the mouse button with the pointer outside of the viewport, it's weird that pressed elements keep the awareness of being pressed.

In other words, I'd suggest modifying the approach a bit to that the "drops" happen if the user eventually releases the mouse button outside of the viewport, as if the viewport was virtually extended only for the sake of dragging, until dragging ends (button released),

**Comment by Meriipu** [comment]
> If the user eventually releases the mouse button with the pointer outside of the viewport, it's weird that pressed elements keep the awareness of being pressed.
> 

By awareness of being pressed, do you mean simply the cosmetic one like scrollbars remaining highlighted?
This happens even if the user releases the mouse button inside the viewport as long as the mouse is not moved after being released (both before and after this pull-request; perhaps by chance the previous behaviour cleared the awareness when the mouse was released outside, but did nothing for when it was released inside). 
This makes me think the awareness remaining after release is a separate issue from the one this pull request is addressing, that could be addressed in a followup. 

I could try to include a change for that too in this, though.

**Comment by Meriipu** [comment]
OK

I still firmly believe that the lingering highlight is a separate issue, but the second commit should fix that (both if the mouse is released but not moved inside the godot window, and if the mouse is released outside the godot window).

**Comment by Meriipu** [comment]
> In other words, I'd suggest modifying the approach a bit to that the "drops" happen if the user eventually releases the mouse button outside of the viewport, as if the viewport was virtually extended only for the sake of dragging, until dragging ends (button released),

it should be done now @RandomShaper

**Comment by RandomShaper** [comment]
Thank you very much for the update. Please give me a few more days to review it. I'm incredibly busy now and I'd like to give your PR the attention it deserves, not just a quick skim through code.

**Comment by RandomShaper** [comment]
I've just checked this. The new changes look cleaner to me.

There's something I'd like to get confirmed: if I understand correctly, this PR now has something I mentioned to be desirable: if you are dragging a scroll bar and then you unpress the button with the pointer outside the window, the scroll thumb loses its hightlight, right?

My only real concern then would be that this PR should only affect GUI, and not physics. Namely, I suggest keeping `_drop_physics_mouseover();` in both cases of the `switch` (probably that involves removing the `[[fallthrough]]` and keeping them as separate cases, even if they need to share a bit of code. Maybe physics can get a similar enhancement in a future, but let's leave them unchanged for now.

**Comment by Meriipu** [comment]
> I've just checked this. The new changes look cleaner to me.
> 
> There's something I'd like to get confirmed: if I understand correctly, this PR now has something I mentioned to be desirable: if you are dragging a scroll bar and then you unpress the button with the pointer outside the window, the scroll thumb loses its hightlight, right?
> 
It loses the highlight if you do that, yes. In addition, with the second commit, it also loses the highlight if the mouse is released (but not moved) inside the window.
Previously, it only lost the highlight if the mouse moved outside the window. 

> My only real concern then would be that this PR should only affect GUI, and not physics. Namely, I suggest keeping `_drop_physics_mouseover();` in both cases of the `switch` (probably that involves removing the `[[fallthrough]]` and keeping them as separate cases, even if they need to share a bit of code. Maybe physics can get a similar enhancement in a future, but let's leave them unchanged for now.

More or less like this? I am not sure what physics_mouseover is, to be honest.
I recall an issue I came across while trying to fix this issue that had a sample project with a square that moved left and right, but I was not able to understand which bug that sample was trying to demonstrate. 

Either way it seems to work for me with this latest version too.
Start godot (project manager) -> Scan 
* Releasing the mouse on the scrollbar drops the drag-highlight and reverts it to its being-hovered highlight
* Moving the mouse outside the window does not drop mouseover
* Releasing the mouse outside the window drops mouseover + reverts highlight to unfocused
* Releasing the mouse inside the window drops mouseover + reverts highlight to unfocused

**Comment by RandomShaper** [comment]
I guess this is good to go. Perhaps there's still some corner case, but this is already a big improvement over what we had formerly.

If anything, please add a counterpart comment to the original of the duplicated code about the duplication, so future maintainers know that those pieces of behavior must match. Refactoring them into one function or something would have been ideal, but I can see it's not easy since the original code has another block of code in between. I guess the comment is good enough.

Finally, I also think should go to both 4.0 and 3.2. I'll tag it accordingly.

File: scene/main/viewport.cpp [hunk 1]
```diff
@@ -1948,6 +1957,33 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
...
 			}*/
 
+			// in case the mouse was released after for example dragging a scrollbar,
+			// check whether the current control is different from the stored one. If
+			// it is different, rather than wait for it to be updated the next time the
+			// mouse is moved, notify the control so that it can e.g. drop the highlight.
+			// This code is duplicated from the mm.is_valid()-case further below.
+			if (gui.mouse_focus) {
+				over = gui.mouse_focus;
+				//recompute focus_inv_xform again here
```
Comment (akien-mga) [inline scene/main/viewport.cpp:1967]: Is this a TODO?

Comment (Meriipu) [reply scene/main/viewport.cpp:1967]: I am not sure. It is exactly the code that is used to update the current hover on mousemove-events further down. 

https://github.com/godotengine/godot/blob/430460b94b76524cd9a563398320e29135dc3acd/scene/main/viewport.cpp#L2040-L2069

Maybe that comment should be removed for the duplicated version.

Comment (Meriipu) [reply scene/main/viewport.cpp:1967]: That comment in the original section that I copied has been around since 0b806ee0fc9097fa7bda7ac0109191c9c5e0a1ac so who knows what the intention behind the comment was.

Comment (RandomShaper) [reply scene/main/viewport.cpp:1967]: Maybe is it time to remove it? An additional benefit would be that the diff will show a change in the original instance of the code, and that would help with traceability.

**Review by RandomShaper — APPROVED** [review]

File: scene/main/viewport.cpp [hunk 2]
```diff
@@ -1955,6 +1957,31 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
 				gui.drag_data=Variant(); //always clear
 			}*/
 
+			// in case the mouse was released after for example dragging a scrollbar,
```
Comment (akien-mga) [inline scene/main/viewport.cpp:1960]: Nitpick, but start comment with an uppercase letter :)

**Comment by Meriipu** [comment]
Is it not so good to rebase and force-push after a review, actually?

File: scene/main/viewport.cpp (hunk 2 above)
Comment (Meriipu) [reply scene/main/viewport.cpp]: There are a lot of existing (full-sentence, not single-word) comments with varying formatting too.

**Comment by akien-mga** [comment]
> Is it not so good to rebase and force-push after a review, actually?

That's our normal workflow :)

File: scene/main/viewport.cpp [hunk 3]
```diff
@@ -1955,6 +1957,31 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
...
+			// it is different, rather than wait for it to be updated the next time the
+			// mouse is moved, notify the control so that it can e.g. drop the highlight.
+			// This code is duplicated from the mm.is_valid()-case further below.
+			if (gui.mouse_focus) {
+				over = gui.mouse_focus;
+			} else {
+				over = _gui_find_control(mpos);
+			}
+
+			if (over != gui.mouse_over) {
```
Comment (reduz) [inline scene/main/viewport.cpp:1971]: I have the feeling this should happen only when gui.mouse_focus_mask == 0 , meaning no mouse button is being pressed, otherwise the logic dictates that the events are captured by the control you are dragging from.

Comment (Meriipu) [reply scene/main/viewport.cpp:1971]: >
>I have the feeling this should happen only when gui.mouse_focus_mask == 0

This way? Still seems to work for me with the extra condition added. Is it to account for the case where multiple buttons might be pressed and one was released?

 @RandomShaper :
>
> If anything, please add a counterpart comment to the original of the duplicated code about the duplication, so future maintainers know that those pieces of behavior must match. 

Sorry, I missed this comment earlier I added a comment above those sections too.

Comment (Meriipu) [reply scene/main/viewport.cpp]: Haha forgot to push.

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.2.4.

**Comment by akien-mga** [comment]
Looks like I missed the first commit of this PR when I cherry-picked back in November... that's why we require that PRs have only a single commit when all the changes it does are related and meant to be treated as a unit.

@RandomShaper Do you think it's safe to cherry-pick 44657db now for 3.2.4 RC 4, given that we have cherry-picked e8804b9 already?

**Comment by akien-mga** [comment]
I went ahead and cherry-picked the commit that I had missed for 3.2.4 RC 4 (via 6f77f1bb3eb8e402b72c678eded585f1a6a870a9). Having only part of the solution would likely cause other issues.

**Comment by RandomShaper** [comment]
99% sure that's fine.

**Comment by Meriipu** [comment]
I interpreted them to be 

> that's why we require that PRs have only a single commit when all the changes it does are related and meant to be treated as a unit.
> 

I think they were related in the sense of both being about focus here but they solved separate issues.

**Comment by akien-mga** [comment]
> I think they were related in the sense of both being about focus here but they solved separate issues.

Then that was fine, it was just my bad for missing the fact that there's two of them, as they're not grouped together in the discussion (PRs with multiple commits usually still have them grouped thanks to successive rebases, but this was not necessary/done here so I missed it. Well, thankfully I was checking other cherry-picks from Nov 11 as I had missed another PR, maybe have been a slow day :P
